### PR TITLE
fix(49869): No error in JavaScript on optional binding pattern parameter

### LINKED ISSUE #49869: No error in JavaScript on optional binding pattern parameter
# Bug Report

### 🔎 Search Terms

TS2463 javascript 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `default`

### ⏯ Playground Link

- [TypeScript link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgPIAczAPYgM7IDeAUAJAC2cAngEYQCqIAJhDKBEwFzJ5hSgBzANzEAvsWIwAriARZcySLwAUJCtTqMWbEB2QBeZACJeUIwBoxAfm4Z5+AJREyYOAGsIeAMp9ByyrQMzKzsTA4i4pIycjggiu6ePvwgAsqm3KaCTmoIuHjYADYQAHQF2Kmm4WISSmDK4UA)
- [JavaScript link](https://www.typescriptlang.org/play?strictNullChecks=true&noUnusedLocals=true&noUnusedParameters=true&noFallthroughCasesInSwitch=true&ts=4.0.5&filetype=js&noUncheckedIndexedAccess=true&exactOptionalPropertyTypes=true&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true&useUnknownInCatchVariables=true#code/PQKhCgAIUgBAXAngBwKYBNUDNIHlnwCWA9gHYDOUMsyATsWrUpAN7ny2GkDmAvpAFsAhogBGqAKqlMWLhirBw4UBGhxkQ2kIGt8RMuX4BtBvooBdBeCwBXUgGMzkeKnYAKFuACQwsZOnYcuiQALyQAETstOEANOC8AJSs3vBCANauAMocXNxuvuJSMkEJANzxSipU6prarFG5-FFWtg5OqRnk2Zw8blFJnl72BsQANqgAdKPEef3lvEou7mVAA)

### 💻 Code

```ts
/**
 * @param [options]
 */
function test({ x = '' }) { x }
```

### 🙁 Actual behavior

No error on `{ x = '' }`.

### 🙂 Expected behavior

Error 2463: "A binding pattern parameter cannot be optional in an implementation signature."

The equivalent Typescript has the correct error:

```ts
function test({ x = '' }?) { x }
```

Note that in the playground link I have checked all options that are supposed to make types more strict. Also note that when this code is run, you'll end up with a runtime error. The same is true for the TypeScript equivalent, but in that case you get an error in advance warning you that the code you wrote will cause issues.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #49869

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -35230,7 +35230,7 @@ namespace ts {
                     error(node.name, Diagnostics.constructor_cannot_be_used_as_a_parameter_property_name);
                 }
             }
-            if (node.questionToken && isBindingPattern(node.name) && (func as FunctionLikeDeclaration).body) {
+            if ((node.questionToken || isJSDocOptionalParameter(node)) && isBindingPattern(node.name) && (func as FunctionLikeDeclaration).body) {
```
Comment (sandersn) [inline src/compiler/checker.ts:35233]: aha! I think this is what I was thinking of in the exactOptionalProperties PR.

**Review by sandersn — APPROVED** [review]

**Comment by voxpelli** [comment]
I think this caused a regression: https://github.com/voxpelli/pony-cause/issues/45

I’ll check it better when I’m at a computer and will try to find time to post a proper issue about it.

**Comment by sandersn** [comment]
@a-tarasyuk looking at the failure, it's a question of whether we want to make Javascript syntax match Typescript syntax: 
https://github.com/voxpelli/pony-cause/issues/45#issuecomment-1201486508

**Comment by sandersn** [comment]
After thinking about it a bit, I think consistency with Typescript is more valuable than being able to be explicit about optionality of binding patterns in JSDoc. So currently I'd recommend leaving this fix as-is and changing pony-cause.

**Comment by voxpelli** [comment]
I disagree, but responded in https://github.com/voxpelli/pony-cause/issues/45#issuecomment-1201572468 rather than here so I won't repeat myself here :)

**Comment by voxpelli** [comment]
To ensure that this doesn't get lost, I added a proper regression report to the TS-repo: https://github.com/microsoft/TypeScript/issues/50286

**Pull request merged**